crypto/merkletrie/trie.go

```diff
@@ -0,0 +1,282 @@
+	if mt.root == storedNodeIdentifierNull {
+		// first item added to the tree.
+		var pnode *node
+		mt.cache.beginTransaction()
+		pnode, mt.root = mt.cache.allocateNewNode()
+		mt.cache.commitTransaction()
+		pnode.leaf = true
+		pnode.hash = d
+		mt.elementLength = len(d)
+		return true, nil
+	}
+	if len(d) != mt.elementLength {
+		return false, ErrMismatchingElementLength
+	}
+	pnode, err := mt.cache.getNode(mt.root)
+	if err != nil {
+		return false, err
+	}
+	found, err := pnode.find(mt.cache, d[:])
+	if found || (err != nil) {
+		return false, err
+	}
+	mt.cache.beginTransaction()
+	var updatedRoot storedNodeIdentifier
+	updatedRoot, err = pnode.add(mt.cache, d[:], make([]byte, 0, len(d)))
+	if err != nil {
+		mt.cache.rollbackTransaction()
+		return false, err
+	}
+	mt.cache.deleteNode(mt.root)
+	mt.root = updatedRoot
+	mt.cache.commitTransaction()
+	return true, nil
+}
+
+// Delete deletes the given hash to the trie, if such element exists.
+// if no such element exists, return false
+func (mt *Trie) Delete(d []byte) (bool, error) {
+	if mt.root == storedNodeIdentifierNull {
+		return false, nil
+	}
+	if len(d) != mt.elementLength {
+		return false, ErrMismatchingElementLength
+	}
+	pnode, err := mt.cache.getNode(mt.root)
+	if err != nil {
+		return false, err
+	}
+	found, err := pnode.find(mt.cache, d[:])
+	if !found || err != nil {
+		return false, err
+	}
+	mt.cache.beginTransaction()
+	if pnode.leaf {
+		// remove the root.
+		mt.cache.deleteNode(mt.root)
+		mt.root = storedNodeIdentifierNull
+		mt.cache.commitTransaction()
+		mt.elementLength = 0
+		return true, nil
+	}
+	var updatedRoot storedNodeIdentifier
+	updatedRoot, err = pnode.remove(mt.cache, d[:], make([]byte, 0, len(d)))
+	if err != nil {
+		mt.cache.rollbackTransaction()
+		return false, err
+	}
+	mt.cache.deleteNode(mt.root)
+	mt.cache.commitTransaction()
+	mt.root = updatedRoot
+	return true, nil
+}
+
+// GetStats return statistics about the merkle trie
+func (mt *Trie) GetStats() (stats Stats, err error) {
+	if mt.root == storedNodeIdentifierNull {
+		return Stats{}, nil
+	}
+	pnode, err := mt.cache.getNode(mt.root)
+	if err != nil {
+		return Stats{}, err
+	}
+	err = pnode.stats(mt.cache, &stats, 1)
+	return
+}
+
+// Commit stores the existings trie using the committer.
+func (mt *Trie) Commit() error {
+	bytes := mt.serialize()
+	mt.cache.committer.StorePage(storedNodeIdentifierNull, bytes)
+	return mt.cache.commit()
+}
+
+// Evict removes elements from the cache that are no longer needed.
+func (mt *Trie) Evict(commit bool) (int, error) {
+	if commit {
+		if mt.cache.modified {
+			if err := mt.Commit(); err != nil {
+				return 0, err
+			}
+		}
+	} else {
+		if mt.cache.modified {
+			return 0, ErrUnableToEvictPendingCommits
+		}
+	}
+	return mt.cache.evict(), nil
+}
+
+// serialize serializes the trie root
+func (mt *Trie) serialize() []byte {
+	serializedBuffer := make([]byte, (8+1)*4) // allocate the worst-case scenario for the trie header.
```

Comment:
I think this should be `(8+1)*5` or there might not be room for `pageSizeLength`.

Reply:
good point.